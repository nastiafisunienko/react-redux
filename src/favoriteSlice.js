import { createSlice } from "@reduxjs/toolkit";

const initialState = {
    favorites: 0,
    items : []
}

const favorite = createSlice({
    name: "favorite",
    initialState,

    reducers: {
        addFavorite : (state, action) => {
            let item = action.payload;

            const existe = state.items.find((elem) => elem.id === item.id)

            if (!existe) {
                state.items.push(item)
                state.favorites++
            }
        },

        deleteFavorite : (state, action) => {
            let item = action.payload.id
            state.items = state.items.filter((elem) => elem.id !== item)
            state.favorites--
        }
    }

})

export const {addFavorite, deleteFavorite} = favorite.actions
export default favorite.reducer