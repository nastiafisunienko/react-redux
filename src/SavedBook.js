import "./Saved.css"
import FavoriteIcon from '@mui/icons-material/Favorite';
import DeleteIcon from '@mui/icons-material/Delete';
import { deleteFavorite } from "./favoriteSlice";
import { useDispatch } from "react-redux";

export default function SavedBook({title, desc, src, author, id}) {
    const dispatch = useDispatch()
    return (
        <div className="bookStyle">
            <FavoriteIcon className="icon"/>
            <h1>{title}</h1>
            <h3>{author}</h3>

            <div>
                <p dangerouslySetInnerHTML={{ __html: desc }}></p>
                <img style={{maxWidth: "50%"}}  src={src} alt="id1"/>
            </div>
            <div className="botones-book">
                <button onClick={() => dispatch(deleteFavorite({id}))}><DeleteIcon className="delete"/></button>

            </div>
        </div>
    )
}