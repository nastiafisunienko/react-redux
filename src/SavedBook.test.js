import { fireEvent, render, screen } from '@testing-library/react';
import SavedBook from './SavedBook';
import { Provider } from 'react-redux';
import { configureStore, createSlice } from '@reduxjs/toolkit';
import { deleteFavorite } from './favoriteSlice';

const mockReducer = createSlice({
    name: "favorites",
    initialState: [],
    reducers: {
        deleteFavorite: (state, action) => {
            return state.filter(fav => fav.id !== action.payload.id);
        }
    }
});

const store = configureStore({
    reducer: {
        favorite: mockReducer.reducer
    }
});

describe("SavedBook component", () => {

    it("makes a snapshot", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <SavedBook title="Book Title" desc="Description" src="image.jpg" author="Author" id="1"/>
            </Provider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("button delete with id works", () => {
        const mockDispatch = jest.fn();
        store.dispatch = mockDispatch;

        render(
            <Provider store={store}>
                <SavedBook title="Book Title" desc="Description" src="image.jpg" author="Author" id="1"/>
            </Provider>
        );

        const button = screen.getByRole("button"); 
     
        expect(button).toBeInTheDocument(); 

        fireEvent.click(button);

        expect(mockDispatch).toHaveBeenCalledWith(deleteFavorite({ id: "1" }));
    });
});