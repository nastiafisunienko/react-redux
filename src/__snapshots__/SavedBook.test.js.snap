// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavedBook component makes a snapshot 1`] = `
<DocumentFragment>
  <div
    class="bookStyle"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-1umw9bq-MuiSvgIcon-root"
      data-testid="FavoriteIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54z"
      />
    </svg>
    <h1>
      Book Title
    </h1>
    <h3>
      Author
    </h3>
    <div>
      <p>
        Description
      </p>
      <img
        alt="id1"
        src="image.jpg"
        style="max-width: 50%;"
      />
    </div>
    <div
      class="botones-book"
    >
      <button>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium delete css-1umw9bq-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`el componente SavedBook makes a snapshot 1`] = `
<DocumentFragment>
  <div
    class="bookStyle"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-1umw9bq-MuiSvgIcon-root"
      data-testid="FavoriteIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54z"
      />
    </svg>
    <h1 />
    <h3 />
    <div>
      <p />
      <img
        alt="id1"
        style="max-width: 50%;"
      />
    </div>
    <div
      class="botones-book"
    >
      <button>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium delete css-1umw9bq-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
